// @universityofmaryland/web-components-library/dist/index-CoevyLaX.js
import { tabs as n } from "@universityofmaryland/web-elements-library/composite";
import { createSlot as c } from "@universityofmaryland/web-utilities-library/elements";
import { Attributes as o, Register as l, Lifecycle as d, Slots as r } from "@universityofmaryland/web-model-library";
const u = "umd-element-tabs", p = {
  tabs: {
    required: !0
  }
}, b = o.handler.common.resize(
  (t) => t.events?.resize()
), f = (t) => {
  const a = o.isTheme.dark({ element: t }), i = c(r.name.TABS), e = t.querySelector(`[slot="${r.name.TABS}"]`);
  return (() => {
    if (!e) return;
    const m = n.elements.DomStrcuture.ModifyElement({
      markup: e
    });
    e.innerHTML = "", e.appendChild(m);
  })(), n.standard({
    isThemeDark: a,
    tabsContainer: e?.children[0],
    shadowContent: i,
    topPosition: o.getValue.topPosition({
      element: t
    })
  });
}, s = l.webComponent({
  tagName: u,
  slots: p,
  createComponent: f,
  attributes: [b],
  afterConnect: d.hooks.loadOnConnect
}), C = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  TabDisplay: s,
  display: s
}, Symbol.toStringTag, { value: "Module" }));
export {
  s as T,
  C as t
};
