// @universityofmaryland/web-components-library/dist/index-BIkAyrlD.js
import { carousel as c } from "@universityofmaryland/web-elements-library/composite";
import { createSlot as I } from "@universityofmaryland/web-utilities-library/elements";
import { Attributes as r, Register as u, Lifecycle as p, Slots as o } from "@universityofmaryland/web-model-library";
import { imageHasAlt as g } from "@universityofmaryland/web-utilities-library/accessibility";
const B = "umd-element-carousel", E = {
  blocks: {
    required: !0
  }
}, D = r.handler.common.resize(
  (e) => e.events?.resize()
), L = (e) => {
  const l = e.shadowRoot, a = r.isTheme.dark({
    element: e
  }), s = e.getAttribute("left-button"), d = e.getAttribute("right-button"), i = e.getAttribute("mobile-hint"), n = e.getAttribute("hint"), t = e.getAttribute("tablet-size"), m = e.getAttribute("desktop-size"), b = e.getAttribute("tablet-count"), f = e.getAttribute("desktop-count"), h = e.getAttribute("max-count"), T = e.getAttribute("grid-gap-pixels"), C = e.querySelector(
    `[slot="${o.name.BLOCKS}"]`
  ), x = Array.from(
    C.querySelectorAll(":scope > *")
  );
  let S = !0, A = !0, $ = !0, k = !0;
  s === "false" && (S = !1), d === "false" && (A = !1), i === "false" && ($ = !1), n === "false" && (k = !1), (() => {
    const O = I(o.name.BLOCKS);
    l.appendChild(O);
  })();
  const R = l.querySelector(
    `[name="${o.name.BLOCKS}"]`
  );
  return c.macro({
    slide: C,
    shadowRef: R,
    blocks: x,
    isThemeDark: a,
    hasLeftButton: S,
    hasRightButton: A,
    mobileHint: $,
    hint: k,
    tabletSize: t ? parseInt(t) : void 0,
    desktopSize: m ? parseInt(m) : void 0,
    tabletCount: b ? parseInt(b) : void 0,
    desktopCount: f ? parseInt(f) : void 0,
    maxCount: h ? parseInt(h) : void 0,
    gridGap: T
  });
}, w = u.webComponent({
  tagName: B,
  slots: E,
  createComponent: L,
  attributes: [D],
  afterConnect: p.hooks.loadOnConnect
}), M = "umd-element-carousel-cards", H = {
  cards: {
    required: !0
  },
  headline: o.element.allowed.headline,
  text: o.element.allowed.text,
  actions: o.element.allowed.actions
}, G = r.handler.common.resize(
  (e) => e.events?.resize()
), F = (e) => {
  const l = e.shadowRoot, a = e.querySelector(
    `[slot="${o.name.CARDS}"]`
  ), s = Array.from(
    a.querySelectorAll(":scope > *")
  );
  (() => {
    const n = I(o.name.CARDS);
    l.appendChild(n);
  })();
  const i = l.querySelector(
    `[name="${o.name.CARDS}"]`
  );
  return c.cards({
    slide: a,
    shadowRef: i,
    cards: s,
    headline: o.headline.default({ element: e }),
    text: o.text.default({ element: e }),
    actions: o.actions.default({ element: e })
  });
}, y = u.webComponent({
  tagName: M,
  slots: H,
  createComponent: F,
  attributes: [G],
  afterConnect: p.hooks.loadOnConnect
}), K = "umd-element-carousel-image", _ = {
  images: {
    required: !0,
    allowedElements: ["img", "div"]
  },
  headlines: {
    allowedElements: ["div"]
  },
  texts: {
    allowedElements: ["div"]
  }
}, j = r.handler.common.resize(
  (e) => e.events?.SetEventReize()
), P = (e) => {
  const l = Array.from(
    e.querySelectorAll(`[slot="${o.name.IMAGES}"] > *`)
  ), a = Array.from(
    e.querySelectorAll(`[slot="${o.name.HEADLINES}"] > *`)
  ), s = Array.from(
    e.querySelectorAll(`[slot="${o.name.TEXTS}"] > *`)
  ), d = a.map(
    (t) => t.cloneNode(!0)
  ), i = s.map(
    (t) => t.cloneNode(!0)
  ), n = l.map((t) => t.nodeName === "IMG" && g(t) ? t.cloneNode(!0) : null).filter((t) => t !== null);
  return c.image({
    images: n,
    headlines: d,
    texts: i,
    isThemeDark: r.isTheme.dark({ element: e }),
    isFullScreenOption: r.includesFeature.fullScreenOption({
      element: e
    })
  });
}, v = u.webComponent({
  tagName: K,
  slots: _,
  createComponent: P,
  attributes: [j]
}), W = "umd-element-carousel-multiple-image", X = {
  images: {
    required: !0,
    allowedElements: ["div", "img"]
  }
}, J = r.handler.common.resize(
  (e) => e.events?.SetEventReize()
), Q = (e) => {
  const a = Array.from(
    e.querySelectorAll(`[slot="${o.name.IMAGES}"] > *`)
  ).map((s) => s.nodeName === "IMG" && g(s) ? s.cloneNode(!0) : null).filter((s) => s !== null);
  return c.images({
    images: a,
    isThemeDark: r.isTheme.dark({ element: e }),
    isFullScreenOption: r.includesFeature.fullScreenOption({
      element: e
    })
  });
}, N = u.webComponent({
  tagName: W,
  slots: X,
  createComponent: Q,
  attributes: [J]
}), U = "umd-element-carousel-thumbnail", V = {
  blocks: {
    required: !0
  }
}, Y = r.handler.common.resize(
  (e) => e.events?.resize()
), Z = (e) => {
  const a = Array.from(
    e.querySelectorAll(`[slot="${o.name.BLOCKS}"] > *`)
  ).map(
    (s) => s.cloneNode(!0)
  );
  return c.thumbnail({
    blocks: a,
    isThemeDark: r.isTheme.dark({ element: e })
  });
}, q = u.webComponent({
  tagName: U,
  slots: V,
  createComponent: Z,
  attributes: [Y],
  afterConnect: p.hooks.loadOnConnect
}), ee = "umd-element-carousel-image-wide", te = {
  slides: {
    required: !0,
    allowedElements: ["figure", "div"]
  }
}, oe = r.handler.common.resize(
  (e) => e.events?.resize()
), se = (e) => {
  const l = e.getAttribute("title"), a = e.getAttribute("animation-time"), s = e.getAttribute("mobile-breakpoint"), i = Array.from(
    e.querySelectorAll('[slot="slides"] > *')
  ).map((n) => {
    const t = n.querySelector("img"), m = n.querySelector("[data-headline]"), b = n.querySelector("[data-text]");
    return t && !g(t) ? (console.warn("Carousel slide image missing alt text:", t), null) : {
      image: t ? t.cloneNode(!0) : null,
      headline: m ? m.cloneNode(!0) : void 0,
      text: b ? b.cloneNode(!0) : void 0
    };
  }).filter((n) => n && n.image);
  return c.wide({
    slides: i,
    title: l || void 0,
    isThemeDark: r.isTheme.dark({ element: e }),
    animationTime: a ? parseInt(a) : void 0,
    mobileBreakpoint: s ? parseInt(s) : void 0
  });
}, z = u.webComponent({
  tagName: ee,
  slots: te,
  createComponent: se,
  attributes: [oe],
  afterConnect: p.hooks.loadOnConnect
}), ce = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  CarouselBase: w,
  CarouselCards: y,
  CarouselImageMultiple: N,
  CarouselImageSingle: v,
  CarouselThumbnail: q,
  CarouselWide: z,
  base: w,
  cards: y,
  imageMultiple: N,
  imageSingle: v,
  thumbnail: q,
  wide: z
}, Symbol.toStringTag, { value: "Module" }));
export {
  w as C,
  y as a,
  N as b,
  v as c,
  q as d,
  z as e,
  ce as f
};
