// @universityofmaryland/web-components-library/dist/index-BXNlNViX.js
import { textLockup as r, events as S } from "@universityofmaryland/web-elements-library/atomic";
import { parseDateFromElement as s, createEventDetails as D } from "@universityofmaryland/web-utilities-library/date";
import { Attributes as i, Register as f, Lifecycle as T, Slots as o } from "@universityofmaryland/web-model-library";
const v = "umd-element-events-date", _ = i.handler.common.resize(
  (e) => e.events?.size()
), b = (e) => {
  const t = i.isTheme.dark({ element: e }), n = o.headline.default({ element: e }), l = e.querySelector(
    `[slot="${o.name.DATE_START_ISO}"]`
  ), m = e.querySelector(
    `[slot="${o.name.DATE_END_ISO}"]`
  ), a = s({
    element: l
  }), d = s({ element: m });
  if (!a)
    return r.date({
      headline: n,
      isThemeDark: t
    });
  const u = D({
    startDate: a,
    endDate: d
  }), p = S.sign({
    ...u,
    isThemeDark: t
  });
  return r.date({
    headline: n,
    dateSign: p,
    isThemeDark: t
  });
}, c = f.webComponent({
  tagName: v,
  createComponent: b,
  attributes: [_],
  afterConnect: T.hooks.loadOnConnect
}), k = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  TextEventLockup: c,
  eventLockup: c
}, Symbol.toStringTag, { value: "Module" }));
export {
  c as T,
  k as t
};
