// @universityofmaryland/web-components-library/dist/index-dODLOINB.js
import { person as e } from "@universityofmaryland/web-elements-library/composite";
import { createStyledSlotOrClone as m, createSlot as c } from "@universityofmaryland/web-utilities-library/elements";
import { Slots as a, Register as d, Attributes as n } from "@universityofmaryland/web-model-library";
import * as C from "@universityofmaryland/web-utilities-library/validation";
const i = ({
  element: o,
  isThemeDark: t
}) => ({
  address: a.contact.address({ element: o }),
  slotTwo: a.person.association({ element: o }),
  name: a.person.name({ element: o }),
  email: a.contact.email({ element: o }),
  slotOne: a.person.jobTitle({ element: o }),
  slotThreeItalic: a.person.pronouns({ element: o }),
  phone: a.contact.phone({ element: o }),
  linkedin: a.social.linkedin({ element: o }),
  bluesky: a.social.bluesky({ element: o }),
  substack: a.social.substack({ element: o }),
  image: C.getValidatedSlotImage({
    element: o,
    slotName: a.name.assets.image
  }),
  additionalContact: a.contact.additional({ element: o }),
  slotFour: a.text.subText({ element: o }),
  actions: a.actions.default({ element: o }),
  isThemeDark: t
}), w = "umd-element-person-bio", E = {
  name: {},
  pronouns: {},
  "job-title": {},
  association: {},
  email: { allowedElements: ["a"] },
  phone: { allowedElements: ["a"] },
  address: {},
  linkedin: { allowedElements: ["a"] },
  "additional-contact": {},
  image: { allowedElements: ["img"] },
  description: {},
  "sub-text": {},
  actions: a.element.allowed.actions
}, D = (o) => {
  const t = n.isTheme.dark({ element: o });
  return n.isLayout.fullImage({ element: o }) ? e.bio.full({
    ...i({ element: o, isThemeDark: t }),
    description: m({
      element: o,
      slotRef: a.name.DESCRIPTION
    })
  }) : e.bio.small({
    ...i({ element: o, isThemeDark: t }),
    description: m({
      element: o,
      slotRef: a.name.DESCRIPTION
    })
  });
}, p = d.webComponent({
  tagName: w,
  slots: E,
  createComponent: D
}), h = "umd-element-person", k = {
  name: {},
  pronouns: {},
  "job-title": {},
  association: {},
  email: { allowedElements: ["a"] },
  phone: { allowedElements: ["a"] },
  address: {},
  linkedin: { allowedElements: ["a"] },
  "additional-contact": {},
  image: { allowedElements: ["img"] },
  "sub-text": {},
  actions: a.element.allowed.actions
}, f = (o) => {
  const t = n.isDisplay.list({ element: o }), l = n.isDisplay.tabular({ element: o }), s = n.isTheme.dark({ element: o });
  return t ? e.list(i({ element: o, isThemeDark: s })) : l ? e.tabular(i({ element: o, isThemeDark: s })) : e.block(i({ element: o, isThemeDark: s }));
}, u = d.webComponent({
  tagName: h,
  slots: k,
  createComponent: f
}), R = "umd-element-person-hero", T = {
  name: {},
  pronouns: {},
  "job-title": {},
  association: {},
  email: { allowedElements: ["a"] },
  phone: { allowedElements: ["a"] },
  address: {},
  linkedin: { allowedElements: ["a"] },
  "additional-contact": {},
  image: { allowedElements: ["img"] },
  "sub-text": {},
  actions: a.element.allowed.actions,
  breadcrumb: { allowedElements: ["umd-element-breadcrumb"] }
}, y = (o) => {
  const t = n.isTheme.dark({
    element: o
  }), l = c(a.name.BREADCRUMB), s = i({ element: o, isThemeDark: t });
  if (l) {
    const r = o.querySelector(
      `[slot="${a.name.BREADCRUMB}"]`
    );
    if (r) {
      const g = r.cloneNode(!0);
      o.appendChild(g), r.setAttribute("slot", a.name.BREADCRUMB_COPY);
    }
  }
  return e.hero({
    ...s,
    pronouns: s.slotThreeItalic,
    breadcrumbDesktop: c(a.name.BREADCRUMB),
    breadcrumbMobile: c(a.name.BREADCRUMB_COPY)
  });
}, b = d.webComponent({
  tagName: R,
  slots: T,
  createComponent: y
}), P = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  PersonBio: p,
  PersonDisplay: u,
  PersonHero: b,
  bio: p,
  display: u,
  hero: b
}, Symbol.toStringTag, { value: "Module" }));
export {
  p as P,
  u as a,
  b,
  P as p
};
