// @universityofmaryland/web-components-library/dist/index-C8NlT4sQ.js
import { pathway as r } from "@universityofmaryland/web-elements-library/composite";
import { toElementVisual as e } from "@universityofmaryland/web-utilities-library/adapters";
import { Register as p, Lifecycle as D, Attributes as i, Slots as t } from "@universityofmaryland/web-model-library";
import { e as T } from "./_event-B6QDfDPW.js";
import { createStyledSlotOrClone as f } from "@universityofmaryland/web-utilities-library/elements";
const x = "umd-element-pathway", m = (a) => {
  const n = a.getAttribute("image-scaled"), u = i.isLayout.imageScaled({ element: a });
  let o = !0;
  u !== null && (o = u), n && (o = n === "true");
  const d = i.isLayout.imagePositionLeft({
    element: a
  }), h = i.isTheme.dark({ element: a }), v = i.isTheme.light({ element: a }), g = i.isTheme.maryland({ element: a }), w = i.isDisplay.hero({ element: a }), b = i.isDisplay.overlay({ element: a }), S = i.isDisplay.sticky({ element: a }), c = i.includesFeature.animation({ element: a }), s = T(a, {
    isLargeSize: !0,
    isDateSignDark: !1
  });
  return w ? r.hero({
    actions: t.actions.default({ element: a }),
    eyebrow: t.eyebrow.default({ element: a }),
    headline: t.headline.default({ element: a }),
    image: t.assets.image({ element: a }),
    includesAnimation: c,
    isImagePositionLeft: d,
    text: t.text.default({ element: a }),
    video: t.assets.video({ element: a })
  }) : b ? r.overlay({
    actions: t.actions.default({ element: a }),
    dateSign: e(s?.dateSign),
    eventDetails: e(s?.eventMeta),
    eyebrow: t.eyebrow.default({ element: a }),
    headline: t.headline.default({ element: a }),
    image: t.assets.image({ element: a }),
    includesAnimation: c,
    isImagePositionLeft: d,
    isImageScaled: o,
    isThemeDark: h,
    isThemeLight: v,
    isThemeMaryland: g,
    stats: t.text.stats({ element: a }),
    text: t.text.default({ element: a }),
    video: t.assets.video({ element: a })
  }) : S ? r.sticky({
    actions: t.actions.default({ element: a }),
    dateSign: e(s?.dateSign),
    eventDetails: e(s?.eventMeta),
    eyebrow: t.eyebrow.default({ element: a }),
    headline: t.headline.default({ element: a }),
    image: t.assets.image({ element: a }),
    isImagePositionLeft: d,
    isImageScaled: o,
    isThemeDark: i.isTheme.dark({ element: a }),
    stats: t.text.stats({ element: a }),
    text: t.text.default({ element: a }),
    video: t.assets.video({ element: a })
  }) : r.standard({
    actions: t.actions.default({ element: a }),
    dateSign: e(s?.dateSign),
    eventDetails: e(s?.eventMeta),
    eyebrow: t.eyebrow.default({ element: a }),
    headline: t.headline.default({ element: a }),
    image: t.assets.image({ element: a }),
    includesAnimation: c,
    isImagePositionLeft: d,
    isImageScaled: o,
    isThemeDark: h,
    isThemeMaryland: g,
    stats: t.text.stats({ element: a }),
    text: t.text.default({ element: a }),
    video: t.assets.video({ element: a })
  });
}, y = p.webComponent({
  tagName: x,
  createComponent: m,
  afterConnect: D.hooks.loadAnimation
}), k = "umd-element-pathway-highlight", I = (a) => r.highlight({
  actions: t.actions.default({ element: a }),
  eyebrow: t.eyebrow.default({ element: a }),
  headline: t.headline.default({ element: a }),
  text: t.text.default({ element: a }),
  quote: f({
    element: a,
    slotRef: t.name.HIGHLIGHT
  }),
  attribution: f({
    element: a,
    slotRef: t.name.HIGHLIGHT_ATTRIBUTION
  }),
  isThemeDark: i.isTheme.dark({ element: a })
}), l = p.webComponent({
  tagName: k,
  createComponent: I
}), M = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  PathwayHighlight: l,
  PathwayImage: y,
  highlight: l,
  image: y
}, Symbol.toStringTag, { value: "Module" }));
export {
  l as P,
  y as a,
  M as p
};
