// @universityofmaryland/web-components-library/dist/index-DJ-AynzD.js
import { navigation as F } from "@universityofmaryland/web-elements-library/composite";
import { Attributes as o, Register as s } from "@universityofmaryland/web-model-library";
import { grid as V, grouped as L, list as $ } from "@universityofmaryland/web-feeds-library/events";
import { bio as x, grid as R, list as M, inTheNews as E } from "@universityofmaryland/web-feeds-library/experts";
import { grid as I, list as S } from "@universityofmaryland/web-feeds-library/in-the-news";
import { featured as O, grid as D, list as z } from "@universityofmaryland/web-feeds-library/news";
const A = "umd-element-navigation-utility", G = (e) => {
  const t = e.hasAttribute("role"), a = e.hasAttribute("aria-label");
  return t || e.setAttribute("role", "navigation"), a || e.setAttribute("aria-label", "Utility navigation"), F.utility({
    alertUrl: o.getValue.alertUrl({ element: e }),
    giftUrl: o.getValue.giftUrl({
      element: e
    }) || "https://giving.umd.edu/giving",
    isAdmissionsFeed: o.isInfo.admissions({ element: e }),
    isAlertOff: o.isLayout.alertOff({ element: e }),
    isEventsFeed: o.isInfo.events({ element: e }),
    isGiftsFeed: o.hasInfo.gifts({ element: e }),
    isLockFull: o.isLayout.lockFull({ element: e }),
    isNewsFeed: o.isInfo.news({ element: e }),
    isSchoolsFeed: o.isInfo.schools({ element: e }),
    isSearch: o.hasInfo.search({ element: e }),
    isSearchDomain: o.isInfo.searchDomain({ element: e })
  });
}, q = o.handler.common.visualShowHide({
  onShow: (e) => e.events?.showAlert(),
  onHide: (e) => e.events?.hideAlert()
}), l = s.webComponent({
  tagName: A,
  createComponent: G,
  attributes: q
}), c = ({ element: e }) => {
  const t = o.getValue.feedToken({ element: e });
  if (!t) {
    console.error("Feed events requires a token to be set");
    return;
  }
  const a = o.getValue.feedFilterIds({ element: e });
  return {
    token: t,
    isThemeDark: o.isTheme.dark({ element: e }),
    isLazyLoad: o.includesFeature.lazyLoad({ element: e }),
    ...a && { categories: a.split(",") }
  };
}, P = "umd-feed-events", U = (e) => {
  const t = c({ element: e });
  if (!t)
    return console.error("Feed events requires a token to be set"), { element: document.createElement("div"), styles: "" };
  const a = Number(o.getValue.layoutColumnCount({ element: e })) || 3, n = Number(o.getValue.layoutRowCount({ element: e })) || 1;
  return V({
    ...t,
    numberOfColumnsToShow: Math.min(Math.max(a, 1), 4),
    numberOfRowsToStart: Math.min(Math.max(n, 1), 2),
    isTransparent: o.isVisual.transparent({ element: e })
  });
}, m = s.webComponent({
  tagName: P,
  createComponent: U,
  afterConnect: (e, t) => {
    e?.events?.callback(t);
  }
}), _ = "umd-feed-events-grouped", j = (e) => {
  const t = c({ element: e });
  if (!t)
    return console.error("Feed events requires a token to be set"), { element: document.createElement("div"), styles: "" };
  const a = Number(o.getValue.layoutRowCount({ element: e })) || 10;
  return L({
    ...t,
    isLazyLoad: !0,
    numberOfRowsToStart: a
  });
}, f = s.webComponent({
  tagName: _,
  createComponent: j,
  afterConnect: (e, t) => {
    e?.events?.callback(t);
  }
}), B = "umd-feed-events-list", H = (e) => {
  const t = c({ element: e });
  if (!t)
    return console.error("Feed events requires a token to be set"), { element: document.createElement("div"), styles: "" };
  const a = Number(o.getValue.layoutRowCount({ element: e })) || 5;
  return $({
    ...t,
    numberOfRowsToStart: Math.min(Math.max(a, 1), 10)
  });
}, p = s.webComponent({
  tagName: B,
  createComponent: H,
  afterConnect: (e, t) => {
    e?.events?.callback(t);
  }
}), J = "umd-feed-expert-bio", K = (e) => {
  const t = o.getValue.feedToken({ element: e }), a = o.getValue.id({ element: e }) || "";
  return !t || !a ? { element: document.createElement("div"), styles: "" } : x({
    token: t,
    expertId: a,
    isThemeDark: o.isTheme.dark({ element: e })
  });
}, g = s.webComponent({
  tagName: J,
  createComponent: K,
  afterConnect: (e, t) => {
    e?.events?.callback(t);
  }
}), Q = "umd-feed-experts-grid", W = (e) => {
  const t = o.getValue.feedToken({ element: e });
  if (!t)
    return { element: document.createElement("div"), styles: "" };
  const a = Number(o.getValue.layoutColumnCount({ element: e })) || 3, n = Number(o.getValue.layoutRowCount({ element: e })) || 1, r = o.getValue.feedFilterIds({ element: e }), i = o.getValue.mediaTrained({ element: e }), u = o.getValue.ids({ element: e }) || "";
  return R({
    token: t,
    numberOfColumnsToShow: Math.min(Math.max(a, 1), 4),
    numberOfRowsToStart: Math.min(Math.max(n, 1), 2),
    isThemeDark: o.isTheme.dark({ element: e }),
    isLazyLoad: o.includesFeature.lazyLoad({ element: e }),
    isTransparent: o.isVisual.transparent({ element: e }),
    isOverlay: o.isDisplay.overlay({ element: e }),
    ...r && { categories: r.split(",") },
    ...i !== null && {
      isMediaTrained: i === "true"
    },
    ...u && { ids: u.split(",") }
  });
}, b = s.webComponent({
  tagName: Q,
  createComponent: W,
  afterConnect: (e, t) => {
    e?.events?.callback(t);
  }
}), X = "umd-feed-experts-list", Y = (e) => {
  const t = o.getValue.feedToken({ element: e });
  if (!t)
    return { element: document.createElement("div"), styles: "" };
  const a = Number(o.getValue.layoutRowCount({ element: e })) || 5, n = o.isDisplay.list({ element: e }), r = o.getValue.feedFilterIds({ element: e }), i = o.getValue.mediaTrained({ element: e }), u = o.getValue.ids({ element: e }) || "";
  return M({
    token: t,
    numberOfRowsToStart: Math.min(Math.max(a, 1), 10),
    cardType: n ? "list" : "tabular",
    isThemeDark: o.isTheme.dark({ element: e }),
    isLazyLoad: o.includesFeature.lazyLoad({ element: e }),
    ...r && { categories: r.split(",") },
    ...i !== null && {
      isMediaTrained: i === "true"
    },
    ...u && { ids: u.split(",") }
  });
}, C = s.webComponent({
  tagName: X,
  createComponent: Y,
  afterConnect: (e, t) => {
    e?.events?.callback(t);
  }
}), Z = "umd-feed-expert-in-the-news", ee = (e) => {
  const t = o.getValue.feedToken({ element: e }), a = o.getValue.id({ element: e }) || "";
  if (!t || !a)
    return console.error(
      "Feed expert in the news requires both data-token and data-id attributes"
    ), { element: document.createElement("div"), styles: "" };
  const n = Number(o.getValue.layoutRowCount({ element: e })) || 5;
  return E({
    token: t,
    expertId: a,
    numberOfRowsToStart: Math.min(Math.max(n, 2), 10),
    isThemeDark: o.isTheme.dark({ element: e }),
    isLazyLoad: o.includesFeature.lazyLoad({ element: e })
  });
}, w = s.webComponent(
  {
    tagName: Z,
    createComponent: ee,
    afterConnect: (e, t) => {
      e?.events?.callback(t);
    }
  }
), N = ({
  element: e
}) => {
  const t = o.getValue.feedToken({ element: e });
  if (!t) {
    console.error("Feed in-the-news requires a token to be set");
    return;
  }
  const a = o.getValue.feedEntryRemoveIds({ element: e });
  return {
    token: t,
    isThemeDark: o.isTheme.dark({ element: e }),
    isLazyLoad: o.includesFeature.lazyLoad({ element: e }),
    ...a && { entriesToRemove: a.split(",") }
  };
}, te = "umd-feed-in-the-news-grid", oe = (e) => {
  const t = N({ element: e });
  if (!t)
    return { element: document.createElement("div"), styles: "" };
  const a = Number(o.getValue.layoutColumnCount({ element: e })) || 3, n = Number(o.getValue.layoutRowCount({ element: e })) || 1;
  return I({
    ...t,
    numberOfColumnsToShow: Math.min(Math.max(a, 1), 4),
    numberOfRowsToStart: Math.min(Math.max(n, 1), 10),
    isTransparent: o.isVisual.transparent({ element: e })
  });
}, h = s.webComponent({
  tagName: te,
  createComponent: oe,
  afterConnect: (e, t) => {
    e?.events?.callback(t);
  }
}), ae = "umd-feed-in-the-news-list", ne = (e) => {
  const t = N({ element: e });
  if (!t)
    return { element: document.createElement("div"), styles: "" };
  const a = Number(o.getValue.layoutRowCount({ element: e })) || 5;
  return S({
    ...t,
    numberOfRowsToStart: a >= 1 && a <= 10 ? a : 5
  });
}, v = s.webComponent({
  tagName: ae,
  createComponent: ne,
  afterConnect: (e, t) => {
    e?.events?.callback(t);
  }
}), d = ({ element: e }) => {
  const t = o.getValue.feedToken({ element: e });
  if (!t) {
    console.error("Feed news requires a token to be set");
    return;
  }
  const a = o.getValue.feedFilterIds({ element: e }), n = o.getValue.feedEntryRemoveIds({ element: e });
  return {
    token: t,
    isThemeDark: o.isTheme.dark({ element: e }),
    isLazyLoad: o.includesFeature.lazyLoad({ element: e }),
    isUnion: o.isData.union({ element: e }),
    ...a && { categories: a.split(",") },
    ...n && { entriesToRemove: n.split(",") }
  };
}, se = "umd-feed-news-featured", re = o.handler.combine(
  o.handler.observe.visuallyPosition({
    callback: (e, t) => e.events?.setPosition(t)
  })
), ie = (e) => {
  const t = d({ element: e });
  if (!t)
    return { element: document.createElement("div"), styles: "" };
  const a = o.getValue.topPosition({ element: e });
  return O({
    ...t,
    numberOfRowsToStart: 1,
    isTransparent: o.isVisual.transparent({ element: e }),
    isLayoutReversed: o.isLayout.reverse({ element: e }),
    overwriteStickyPosition: a ? parseInt(a) : void 0
  });
}, y = s.webComponent({
  tagName: se,
  attributes: re,
  createComponent: ie,
  afterConnect: (e, t) => {
    e?.events?.callback(t);
  }
}), ue = "umd-feed-news", ce = (e) => {
  const t = d({ element: e });
  if (!t)
    return { element: document.createElement("div"), styles: "" };
  const a = Number(o.getValue.layoutColumnCount({ element: e })) || 3, n = Number(o.getValue.layoutRowCount({ element: e })) || 1;
  return D({
    ...t,
    numberOfColumnsToShow: Math.min(Math.max(a, 1), 4),
    numberOfRowsToStart: Math.min(Math.max(n, 1), 2),
    isTransparent: o.isVisual.transparent({ element: e }),
    isTypeOverlay: o.isDisplay.overlay({ element: e })
  });
}, T = s.webComponent({
  tagName: ue,
  createComponent: ce,
  afterConnect: (e, t) => {
    e?.events?.callback(t);
  }
}), de = "umd-feed-news-list", le = (e) => {
  const t = d({ element: e });
  if (!t)
    return { element: document.createElement("div"), styles: "" };
  const a = Number(o.getValue.layoutRowCount({ element: e })) || 5;
  return z({
    ...t,
    numberOfRowsToStart: a >= 1 && a <= 10 ? a : 5
  });
}, k = s.webComponent({
  tagName: de,
  createComponent: le,
  afterConnect: (e, t) => {
    e?.events?.callback(t);
  }
}), we = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  FeedAlert: l,
  FeedEventsGrid: m,
  FeedEventsGrouped: f,
  FeedEventsList: p,
  FeedExpertInTheNews: w,
  FeedExpertsBio: g,
  FeedExpertsGrid: b,
  FeedExpertsList: C,
  FeedInTheNewsGrid: h,
  FeedInTheNewsList: v,
  FeedNewsFeatured: y,
  FeedNewsGrid: T,
  FeedNewsList: k,
  alert: l,
  eventsGrid: m,
  eventsGrouped: f,
  eventsList: p,
  expertsBio: g,
  expertsGrid: b,
  expertsInTheNews: w,
  expertsList: C,
  inTheNewsGrid: h,
  inTheNewsList: v,
  newsFeatured: y,
  newsGrid: T,
  newsList: k
}, Symbol.toStringTag, { value: "Module" }));
export {
  l as F,
  m as a,
  f as b,
  p as c,
  g as d,
  b as e,
  C as f,
  y as g,
  T as h,
  k as i,
  we as j,
  w as k,
  h as l,
  v as m
};
