// @universityofmaryland/web-components-library/dist/index-BrpAs-2C.js
import { slider as m } from "@universityofmaryland/web-elements-library/composite";
import { Attributes as r, Register as c, Lifecycle as u, Slots as n } from "@universityofmaryland/web-model-library";
import { slider as f } from "@universityofmaryland/web-feeds-library/academic";
import { slider as p } from "@universityofmaryland/web-feeds-library/events";
const T = "umd-element-slider-events", v = r.handler.common.resize(
  (e) => e.events?.size()
), S = (e) => {
  const t = document.createElement("div"), o = e.querySelector(
    `[slot=${n.name.EVENT_LIST}]`
  );
  return o ? (t.innerHTML = o.innerHTML, m.events({
    isThemeDark: r.isTheme.dark({ element: e }),
    dataSlider: t,
    headline: n.headline.default({ element: e }),
    actions: n.actions.default({ element: e })
  })) : (console.error(`Slot ${n.name.EVENT_LIST} is required`), { element: t, styles: "" });
}, a = c.webComponent({
  tagName: T,
  createComponent: S,
  attributes: [v],
  afterConnect: u.hooks.loadOnConnect
}), h = "umd-element-slider-events-feed", b = r.handler.common.resize(
  (e) => e.events?.size()
), y = (e) => {
  const t = r.isTheme.dark({ element: e }), o = r.getValue.feedToken({ element: e }), d = r.getValue.feedFilterIds({ element: e }), l = r.isData.type.academic({
    element: e
  });
  if (!o)
    return console.error("Token is required for this component"), { element: document.createElement("div"), styles: "" };
  const s = {
    token: o,
    categories: d,
    isThemeDark: t,
    headline: n.headline.default({ element: e }),
    actions: n.actions.default({ element: e })
  };
  return l ? f(s) : p(s);
}, i = c.webComponent({
  tagName: h,
  createComponent: y,
  attributes: [b],
  afterConnect: (e, t) => e?.events?.callback?.(t)
}), $ = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  SliderEventDisplay: a,
  SliderEventFeed: i,
  eventDisplay: a,
  eventFeed: i
}, Symbol.toStringTag, { value: "Module" }));
export {
  a as S,
  i as a,
  $ as s
};
