// @universityofmaryland/web-components-library/dist/index-CwiTV8_N.js
import { alert as d, banner as i } from "@universityofmaryland/web-elements-library/composite";
import { Slots as e, Register as a, Attributes as o } from "@universityofmaryland/web-model-library";
const c = "umd-element-alert-page", r = {
  headline: e.element.allowed.headline,
  body: e.element.allowed.body,
  text: e.element.allowed.text,
  actions: e.element.allowed.actions
}, m = (t) => d.page({
  headline: e.headline.default({ element: t }),
  text: e.deprecated.body({ element: t }) || e.text.default({ element: t }),
  actions: e.actions.default({ element: t }),
  isThemeLight: o.isTheme.light({ element: t }),
  isThemeDark: o.isTheme.dark({ element: t }),
  isShowIcon: o.isVisual.showIcon({ element: t })
}), l = a.webComponent({
  tagName: c,
  slots: r,
  createComponent: m
}), h = "umd-element-banner-promo", p = {
  headline: e.element.allowed.headline,
  body: e.element.allowed.body,
  text: e.element.allowed.text,
  actions: e.element.allowed.actions
}, u = (t) => i.promo({
  headline: e.headline.default({ element: t }),
  text: e.deprecated.body({ element: t }) || e.text.default({ element: t }),
  actions: e.actions.default({ element: t }),
  isThemeDark: o.isTheme.dark({ element: t }),
  includeSeal: o.isVisual.icon_seal({ element: t })
}), n = a.webComponent({
  tagName: h,
  slots: p,
  createComponent: u
}), b = "umd-element-alert-site", w = {
  headline: e.element.allowed.headline,
  body: e.element.allowed.body,
  text: e.element.allowed.text,
  actions: e.element.allowed.actions
}, g = (t) => d.site({
  headline: e.headline.default({ element: t }),
  text: e.deprecated.body({ element: t }) || e.text.default({ element: t }),
  actions: e.actions.default({ element: t }),
  daysToHide: o.getValue.daysToHide({ element: t }) || "10"
}), s = a.webComponent({
  tagName: b,
  slots: w,
  createComponent: g
}), y = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  AlertPage: l,
  AlertPromo: n,
  AlertSite: s,
  page: l,
  promo: n,
  site: s
}, Symbol.toStringTag, { value: "Module" }));
export {
  l as A,
  n as a,
  s as b,
  y as c
};
