// @universityofmaryland/web-components-library/dist/index-DvXDxhj1.js
import { accordion as a } from "@universityofmaryland/web-elements-library/composite";
import { Slots as t, Attributes as o, Register as r } from "@universityofmaryland/web-model-library";
const n = "umd-element-accordion-item", d = (e) => a.item({
  text: t.deprecated.body({ element: e, isDefaultStyling: !1 }) || t.text.default({ element: e, isDefaultStyling: !1 }),
  headline: t.headline.default({ element: e }),
  isThemeLight: o.isTheme.light({
    element: e
  }),
  isThemeDark: o.isTheme.dark({
    element: e
  }),
  isStateOpen: o.isVisual.open({
    element: e
  })
}), l = {
  headline: {
    ...t.element.allowed.headline,
    required: !0
  },
  body: t.element.allowed.body,
  text: {
    ...t.element.allowed.text,
    required: !0
  }
}, s = o.handler.common.accordion(), i = r.webComponent({
  tagName: n,
  slots: l,
  createComponent: d,
  attributes: s
}), u = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  AccordionItem: i,
  item: i
}, Symbol.toStringTag, { value: "Module" }));
export {
  i as A,
  u as a
};
