// @universityofmaryland/web-components-library/dist/index-DOqPBmZz.js
import { media as n } from "@universityofmaryland/web-elements-library/composite";
import * as u from "@universityofmaryland/web-utilities-library/validation";
import { Slots as t, Register as r, Lifecycle as d, Attributes as s } from "@universityofmaryland/web-model-library";
const w = "umd-element-media-inline", x = {
  caption: {
    allowedElements: ["div", "p"]
  },
  text: t.element.allowed.text,
  wrappingText: {
    allowedElements: ["div", "p"],
    deprecated: 'Use "text" instead. This attribute will be removed in version 2.0.'
  }
}, h = (e) => {
  const i = t.text.caption({ element: e }), o = t.text.default({ element: e, isDefaultStyling: !1 }) || t.deprecated.wrappingText({ element: e, isDefaultStyling: !1 }), c = s.isLayout.alignmentRight({ element: e }), p = s.isTheme.dark({ element: e }), g = o !== null, f = i !== null, a = {
    isAlignmentRight: c,
    isThemeDark: p,
    caption: i,
    image: u.getValidatedSlotImage({
      element: e,
      slotName: t.name.assets.image
    }),
    wrappingText: o
  };
  return g ? n.inline.wrapped(a) : f ? n.inline.caption(a) : n.inline.standard(a);
}, l = r.webComponent({
  tagName: w,
  slots: x,
  createComponent: h,
  onReady: d.hooks.loadOnConnect
}), b = "umd-element-media-gif", y = {
  image: {
    allowedElements: ["img", "a"]
  }
}, C = (e) => n.elements.gif({
  image: t.assets.image({ element: e })
}), m = r.webComponent({
  tagName: b,
  slots: y,
  createComponent: C,
  onReady: d.hooks.loadOnConnect
}), S = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  MediaGif: m,
  MediaInline: l,
  gif: m,
  inline: l
}, Symbol.toStringTag, { value: "Module" }));
export {
  m as M,
  l as a,
  S as m
};
