// @universityofmaryland/web-components-library/dist/index-kYufzuXV.js
import { footer as d } from "@universityofmaryland/web-elements-library/composite";
import { Register as f, Attributes as l, Slots as s } from "@universityofmaryland/web-model-library";
const C = "umd-element-footer", t = {
  CONTACT_HEADLINE: "contact-headline",
  CONTACT_LINKS: "contact-links",
  SOCIAL: "social-links",
  UTILITY: "utility-links",
  LINK_COLUMN_ONE: "link-column-one",
  LINK_COLUMN_TWO: "link-column-two",
  LINK_COLUMN_THREE: "link-column-three",
  LOGO: "logo"
}, h = {
  image: {
    allowedElements: ["img"]
  }
}, S = (o) => {
  const r = l.isDisplay.mega({ element: o }), e = l.isDisplay.visual({ element: o }), N = l.isVisual.campaign({ element: o }), a = o.querySelector(
    `[slot="${t.LINK_COLUMN_ONE}"]`
  ), c = o.querySelector(
    `[slot="${t.LINK_COLUMN_TWO}"]`
  ), n = o.querySelector(
    `[slot="${t.LINK_COLUMN_THREE}"]`
  ), O = o.hasAttribute("role"), T = o.hasAttribute("aria-label"), i = [];
  let g = l.isDisplay.simple({ element: o });
  const u = o.querySelector(`[slot="${t.LOGO}"]`);
  u && u instanceof HTMLAnchorElement && (u.querySelector("img") || console.error("No logo element found in logo slot")), a && a.hasChildNodes() && i.push(a), c && c.hasChildNodes() && i.push(c), n && n.hasChildNodes() && i.push(n), !r && !e && (g = !0), O || o.setAttribute("role", "contentinfo"), T || o.setAttribute("aria-label", "site footer");
  const L = {
    isThemeLight: l.isTheme.light({ element: o }),
    isTypeSimple: g,
    isTypeVisual: e,
    isTypeMega: r,
    isCampaignForward: N,
    slotVisualImage: s.assets.image({ element: o }) || s.deprecated.backgroundImage({ element: o }),
    slotUtilityLinks: o.querySelector(
      `[slot="${t.UTILITY}"]`
    ),
    slotCta: s.actions.default({ element: o }) || s.deprecated.callToAction({ element: o }),
    slotLogo: o.querySelector(`[slot="${t.LOGO}"]`),
    slotAddress: s.contact.address({ element: o }) || s.deprecated.contactAddress({ element: o }),
    slotContentLinks: o.querySelector(
      `[slot="${t.CONTACT_LINKS}"]`
    ),
    slotHeadline: o.querySelector(
      `[slot="${t.CONTACT_HEADLINE}"]`
    ),
    slotSocialLinks: o.querySelector(
      `[slot="${t.SOCIAL}"]`
    ),
    slotColumns: i.length > 0 ? i : void 0
  };
  return r ? d.mega(L) : e ? d.visual(L) : d.simple(L);
}, p = f.webComponent({
  tagName: C,
  slots: h,
  createComponent: S
}), m = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  FooterOptions: p,
  options: p
}, Symbol.toStringTag, { value: "Module" }));
export {
  p as F,
  m as f
};
