// @universityofmaryland/web-components-library/dist/index-ZR5YJ3lO.js
import { hero as n } from "@universityofmaryland/web-elements-library/composite";
import { Register as c, Lifecycle as u, Attributes as t, Slots as o } from "@universityofmaryland/web-model-library";
import { createSlot as h } from "@universityofmaryland/web-utilities-library/elements";
import { getValidatedSlotImage as S } from "@universityofmaryland/web-utilities-library/validation";
const k = "umd-element-hero", x = (e) => {
  const a = ({ element: d }) => ({
    eyebrow: o.eyebrow.default({ element: d }),
    headline: o.headline.default({ element: d }),
    text: o.text.default({ element: d }),
    video: o.assets.video({ element: d }),
    image: o.assets.image({ element: d }),
    actions: o.actions.default({ element: d })
  }), r = t.isDisplay.overlay({ element: e }), l = t.isDisplay.stacked({ element: e }), s = t.includesFeature.animation({ element: e }), i = t.isTheme.dark({ element: e });
  return r ? n.overlay({
    ...a({ element: e }),
    includesAnimation: s
  }) : l ? n.stacked({
    ...a({ element: e }),
    isHeightSmall: t.isLayout.heightSmall({ element: e }) || !1,
    isWidthLarge: t.isLayout.spaceHorizontalLarge({ element: e }) || !1,
    isThemeDark: i,
    includesAnimation: s,
    topPosition: t.getValue.topPosition({
      element: e
    })
  }) : t.isDisplay.stackedInterior({
    element: e
  }) ? n.stacked({
    ...a({ element: e }),
    isHeightSmall: !0,
    isWidthLarge: !0,
    isThemeDark: i,
    includesAnimation: s
  }) : t.isDisplay.standardCentered({
    element: e
  }) ? n.standard({
    ...a({ element: e }),
    isTextCenter: !0,
    isHeightSmall: !1,
    includesAnimation: s
  }) : t.isDisplay.standardInterior({
    element: e
  }) ? n.standard({
    ...a({ element: e }),
    isHeightSmall: !0,
    isThemeDark: i,
    includesAnimation: s
  }) : t.isDisplay.standardInteriorCentered({
    element: e
  }) ? n.standard({
    ...a({ element: e }),
    isHeightSmall: !0,
    isTextCenter: !0,
    includesAnimation: s
  }) : n.standard({
    ...a({ element: e }),
    isHeightSmall: t.isLayout.heightSmall({ element: e }) || !1,
    isTextCenter: t.isLayout.textCentered({ element: e }) || !1,
    isThemeDark: i,
    includesAnimation: s
  });
}, f = c.webComponent({
  tagName: k,
  createComponent: x,
  afterConnect: u.hooks.loadOnConnect
}), L = "umd-element-hero-brand-video", w = (e) => {
  const a = e.getAttribute("animation-trigger"), r = e.querySelector("video");
  return n.custom.videoArrow({
    video: r,
    headline: o.headline.default({ element: e }),
    text: o.text.default({ element: e }),
    isAnimationOnLoad: a === "load"
  });
}, p = c.webComponent({
  tagName: L,
  createComponent: w,
  afterConnect: u.hooks.loadOnConnect
}), A = "umd-element-hero-grid", D = (e) => {
  const a = e.querySelector(
    '[slot="column-left"]'
  ), r = e.querySelector(
    '[slot="column-right"]'
  ), l = e.querySelector(
    '[slot="column-center"]'
  ), s = Array.from(
    a.querySelectorAll("img")
  ), i = Array.from(
    r.querySelectorAll("img")
  ), d = Array.from(
    l.querySelectorAll("img")
  ), m = l.querySelector(
    "video"
  ), b = {
    corners: [
      {
        images: s,
        isCornerLeft: !0
      },
      {
        images: i,
        isCornerLeft: !1
      }
    ],
    center: {
      images: d,
      video: m
    },
    headline: o.headline.default({ element: e }),
    text: o.text.default({ element: e }),
    actions: o.actions.default({ element: e }),
    isThemeDark: t.isTheme.dark({ element: e }) || !1
  }, g = n.custom.grid(b);
  return g || { element: document.createElement("div"), styles: "" };
}, y = c.webComponent({
  tagName: A,
  createComponent: D
}), H = "umd-element-hero-expand", N = t.handler.combine(
  t.handler.observe.visuallyPosition({
    callback: (e, a) => e.events?.setTopPosition({ value: a })
  }),
  // Deprecated
  t.handler.observe.visuallyPosition({
    name: t.names.deprecated.layout.LAYOUT_STICKY_TOP,
    callback: (e, a) => e.events?.setTopPosition({ value: a })
  })
), P = (e) => {
  const a = S({
    element: e,
    slotName: o.name.assets.image
  }), r = e.querySelector('[slot="video"]'), l = h(o.name.actions.default), s = h(o.name.ADDITIONAL), i = {
    eyebrow: o.eyebrow.default({ element: e }),
    headline: o.headline.default({ element: e }),
    topPosition: t.getValue.topPosition({ element: e })
  }, d = r instanceof HTMLVideoElement;
  if (a && (i.image = a), d && (i.video = r), !d && r && r.children.length > 0) {
    const m = r.querySelector("video");
    m && (i.video = m);
  }
  return l && (i.actions = l), s && (i.additional = s), n.custom.expand(i);
}, C = c.webComponent({
  tagName: H,
  createComponent: P,
  attributes: N,
  afterConnect: (e) => {
    if (e.element instanceof HTMLElement) {
      const a = t.getValue.topPosition({
        element: e.element
      });
      a && e?.events?.setTopPosition({ value: a });
    }
  }
}), O = "umd-element-hero-logo", $ = (e) => n.logo({
  isThemeDark: t.isTheme.dark({
    element: e
  }),
  isThemeLight: t.isTheme.light({
    element: e
  }),
  isThemeMaryland: t.isTheme.maryland({
    element: e
  }),
  eyebrow: o.eyebrow.default({ element: e }),
  headline: o.headline.default({ element: e }),
  text: o.text.default({ element: e }),
  image: o.assets.image({ element: e }),
  actions: o.actions.default({ element: e })
}), v = c.webComponent({
  tagName: O,
  createComponent: $,
  afterConnect: u.hooks.loadOnConnect
}), q = "umd-element-hero-minimal", I = (e) => n.minimal({
  isThemeDark: t.isTheme.dark({
    element: e
  }),
  isThemeLight: t.isTheme.light({
    element: e
  }),
  isThemeMaryland: t.isTheme.maryland({
    element: e
  }),
  eyebrow: o.eyebrow.default({ element: e }),
  headline: o.headline.default({ element: e }),
  text: o.text.default({ element: e }),
  image: o.assets.image({ element: e }),
  actions: o.actions.default({ element: e })
}), T = c.webComponent({
  tagName: q,
  createComponent: I,
  afterConnect: u.hooks.loadOnConnect
}), j = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  HeroBase: f,
  HeroExpand: C,
  HeroGrid: y,
  HeroLogo: v,
  HeroMinimal: T,
  HeroVideoArrow: p,
  base: f,
  expand: C,
  grid: y,
  logo: v,
  minimal: T,
  videoArrow: p
}, Symbol.toStringTag, { value: "Module" }));
export {
  f as H,
  C as a,
  y as b,
  v as c,
  T as d,
  p as e,
  j as h
};
