// @universityofmaryland/web-components-library/dist/index-BgXjsdgw.js
import { actions as i } from "@universityofmaryland/web-elements-library/atomic";
import { Register as c, Attributes as e, Slots as r } from "@universityofmaryland/web-model-library";
const u = "umd-element-call-to-action", d = {
  text: {
    allowedElements: ["a", "button"]
  },
  plainText: {
    deprecated: 'Use "text" instead. This attribute will be removed in version 2.0.',
    allowedElements: ["a", "button"]
  }
}, y = (t) => {
  const o = t.querySelector("button, a:not([slot])"), n = t.querySelector(
    '[slot="plain-text"]'
  ), l = t.querySelector('[slot="text"]');
  if (!o)
    throw new Error("Component requires a button or link");
  const s = {
    element: o.cloneNode(!0),
    isTypeOutline: e.isDisplay.outline({ element: t }),
    isTypePrimary: e.isDisplay.primary({ element: t }),
    isTypeSecondary: e.isDisplay.secondary({ element: t }),
    isSizeLarge: e.isVisual.sizeLarge({ element: t }),
    isThemeDark: e.isTheme.dark({ element: t }),
    isThemeGold: e.isTheme.gold({ element: t }),
    elementStyles: e.getValue.styleProps({ element: t })
  };
  if (n || l) {
    const p = r.deprecated.plainText({ element: t }) || r.text.default({ element: t });
    return i.options({
      ...s,
      plainText: p
    });
  }
  return i.options({
    ...s
  });
}, a = c.webComponent({
  tagName: u,
  slots: d,
  createComponent: y
}), b = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  ActionsDisplay: a,
  display: a
}, Symbol.toStringTag, { value: "Module" }));
export {
  a as A,
  b as a
};
