// @universityofmaryland/web-components-library/dist/index-LBElx1HN.js
import { layout as s } from "@universityofmaryland/web-elements-library/composite";
import * as g from "@universityofmaryland/web-utilities-library/validation";
import { Register as n, Attributes as e, Slots as a, Lifecycle as b } from "@universityofmaryland/web-model-library";
import { createSlot as i } from "@universityofmaryland/web-utilities-library/elements";
import { layout as h } from "@universityofmaryland/web-elements-library/atomic";
const S = "umd-element-logo", T = {
  image: {
    required: !0
  }
}, k = (t) => s.box.logo({
  image: g.getValidatedSlotImage({
    element: t,
    slotName: a.name.assets.image
  }),
  text: a.text.default({ element: t }),
  isThemeDark: e.isTheme.dark({ element: t }),
  isBordered: e.isVisual.bordered({ element: t })
}), c = n.webComponent({
  tagName: S,
  slots: T,
  createComponent: k
}), L = "umd-layout-image-expand", N = {
  content: {
    required: !0
  },
  image: {
    required: !0
  }
}, f = (t) => s.image.expand({
  content: i("content"),
  image: g.getValidatedSlotImage({
    element: t,
    slotName: a.name.assets.image
  })
}), r = n.webComponent({
  tagName: L,
  slots: N,
  createComponent: f
}), $ = "umd-element-modal", v = (t) => {
  const o = () => {
    t.setAttribute(
      e.names.layout.hidden,
      e.values.state.TRUE
    );
  }, y = {
    content: a.content.default({ element: t, isDefaultStyling: !1 }),
    isHidden: e.isLayout.hidden({ element: t }),
    context: t,
    callback: o
  };
  return h.overlay.modal(y);
}, I = e.handler.common.visualShowHide({
  onShow: (t) => t.events?.show(),
  onHide: (t) => t.events?.hide()
}), m = n.webComponent({
  tagName: $,
  attributes: I,
  createComponent: v
}), x = "umd-element-scroll-top", w = (t) => {
  const o = {
    isFixed: t.hasAttribute(
      e.names.deprecated.layout.LAYOUT_FIXED
    )
  };
  return s.scrollTop(o);
}, l = n.webComponent({
  tagName: x,
  createComponent: w,
  afterConnect: b.hooks.loadOnConnect
}), C = ({
  element: t,
  isThemeDark: o
}) => ({
  headline: a.headline.default({ element: t }),
  actions: a.actions.default({ element: t }),
  isThemeDark: o
}), A = "umd-element-section-intro", O = (t) => s.sectionIntro.small({
  ...C({
    element: t,
    isThemeDark: e.isTheme.dark({ element: t })
  }),
  text: a.text.default({ element: t }),
  hasSeparator: t.hasAttribute(
    e.names.deprecated.option.OPTIONAL_HAS_SEPARATOR
  )
}), u = n.webComponent({
  tagName: A,
  createComponent: O
}), _ = "umd-element-section-intro-wide", P = (t) => s.sectionIntro.wide(
  C({
    element: t,
    isThemeDark: e.isTheme.dark({ element: t })
  })
), d = n.webComponent({
  tagName: _,
  createComponent: P
}), D = "umd-element-sticky-columns", q = {
  "sticky-column": {
    required: !0
  },
  "static-column": {
    required: !0
  }
}, E = e.handler.combine(
  e.handler.observe.visuallyPosition({
    callback: (t, o) => t.events?.setPosition({ value: o })
  }),
  // Deprecated
  e.handler.observe.visuallyPosition({
    name: e.names.deprecated.layout.LAYOUT_STICKY_TOP,
    callback: (t, o) => t.events?.setPosition({ value: o })
  })
), U = (t) => s.stickyColumns({
  stickyColumn: i(a.name.STICKY_COLUMN),
  staticColumn: i(a.name.STATIC_COLUMN),
  isStickyLast: !e.isVisual.stickyFirst({ element: t }),
  topPosition: e.getValue.topPosition({ element: t })
}), p = n.webComponent({
  tagName: D,
  slots: q,
  createComponent: U,
  attributes: E
}), Y = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  LayoutBoxLogo: c,
  LayoutImageExpand: r,
  LayoutModal: m,
  LayoutScrollTop: l,
  LayoutSectionIntroSmall: u,
  LayoutSectionIntroWide: d,
  LayoutStickyColumns: p,
  boxLogo: c,
  imageExpand: r,
  modal: m,
  scrollTop: l,
  sectionIntroSmall: u,
  sectionIntroWide: d,
  stickyColumns: p
}, Symbol.toStringTag, { value: "Module" }));
export {
  c as L,
  r as a,
  m as b,
  l as c,
  u as d,
  d as e,
  p as f,
  Y as l
};
