// @universityofmaryland/web-components-library/dist/index-OptCax2o.js
import { quote as T } from "@universityofmaryland/web-elements-library/composite";
import { createStyledSlotOrClone as i } from "@universityofmaryland/web-utilities-library/elements";
import { isPreferredReducedMotion as y } from "@universityofmaryland/web-utilities-library/accessibility";
import { getLinkFromSlot as S } from "@universityofmaryland/web-utilities-library/dom";
import { Register as b, Lifecycle as q, Attributes as o, Slots as e } from "@universityofmaryland/web-model-library";
const R = "umd-element-quote", A = {
  quote: {
    required: !0
  }
}, O = ({
  element: t
}) => {
  const s = e.actions.default({ element: t });
  return S(s);
}, _ = (t) => {
  const s = o.isVisual.sizeLarge({ element: t }), g = o.isDisplay.featured({ element: t }), r = o.isTheme.dark({ element: t }), n = o.isTheme.maryland({ element: t }), a = o.isVisual.transparent({ element: t }), c = O({ element: t }), u = i({
    element: t,
    slotRef: e.name.QUOTE
  }), m = i({
    element: t,
    slotRef: e.name.assets.image
  }), l = i({
    element: t,
    slotRef: e.name.ATTRIBUTION
  }), d = i({
    element: t,
    slotRef: e.name.ATTRIBUTION_SUB_TEXT
  }), f = o.includesFeature.animation({ element: t }) && !y();
  return g ? T.featured({
    isSizeLarge: s,
    isThemeDark: r,
    isThemeMaryland: n,
    isTransparent: a,
    includesAnimation: f,
    action: c,
    quote: u,
    image: m,
    attribution: l,
    attributionSubText: d
  }) : T.inline({
    isSizeLarge: s,
    isThemeDark: r,
    isThemeMaryland: n,
    isTransparent: a,
    includesAnimation: f,
    action: c,
    quote: u,
    image: m,
    attribution: l,
    attributionSubText: d
  });
}, p = b.webComponent({
  tagName: R,
  slots: A,
  createComponent: _,
  afterConnect: q.hooks.loadAnimation
}), U = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  QuoteDisplay: p,
  display: p
}, Symbol.toStringTag, { value: "Module" }));
export {
  p as Q,
  U as q
};
