// @universityofmaryland/web-components-library/dist/index-BqwGw7fw.js
import { animations as l } from "@universityofmaryland/web-elements-library/atomic";
import { Register as m, Lifecycle as d, Attributes as s } from "@universityofmaryland/web-model-library";
const r = "umd-element-brand-card-stack", u = (n) => {
  const t = Array.from(
    n.querySelectorAll('[slot="cards"] > *')
  ), o = t.find(
    (e) => e.getAttribute("data-featured") === "true"
  );
  if (t.length < 5)
    return console.log(`${r}: The minimum number of for the card stack is 5.`), { element: document.createElement("div"), styles: "" };
  if (t.length > 9)
    return console.log(`${r}: The maximum number of for the card stack is 10.`), { element: document.createElement("div"), styles: "" };
  if (!o)
    return console.log(
      `${r}: A featured element is required. Please add the attribute data-featured to one of the elements`
    ), { element: document.createElement("div"), styles: "" };
  const i = t.filter((e) => e.getAttribute("data-featured") !== "true").map((e) => e.tagName === "IMG" ? e : e.querySelector("img")).filter((e) => e !== null);
  return l.brand.cardStack({
    featured: o,
    images: i,
    isExpandFeature: s.includesFeature.imageExpand({
      element: n
    })
  });
}, a = m.webComponent({
  tagName: r,
  createComponent: u,
  afterConnect: d.hooks.loadAnimation
}), f = "umd-element-brand-logo-animation", g = () => l.brand.chevronScroll(), c = m.webComponent({
  tagName: f,
  createComponent: g
}), C = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  BrandCardStack: a,
  BrandChevronScroll: c,
  cardStack: a,
  chevronScroll: c
}, Symbol.toStringTag, { value: "Module" }));
export {
  a as B,
  c as a,
  C as b
};
