// @universityofmaryland/web-components-library/dist/loader-DDtb0UN_.js
import { A as r, a as t } from "./index-DvXDxhj1.js";
import { A as i, a as n } from "./index-BgXjsdgw.js";
import { A as d, a as m, b as l, c as p } from "./index-CwiTV8_N.js";
import { B as c, a as f, b as u } from "./index-BqwGw7fw.js";
import { C as y, a as C, b, c as S, d as g, e as v, f as L } from "./index-ChvgI4-j.js";
import { C as E, a as F, b as h, c as P, d as A, e as w, f as B } from "./index-BIkAyrlD.js";
import { F as H, a as I, b as N, c as x, d as D, e as O, f as T, g as G, h as j, i as k, j as M } from "./index-DJ-AynzD.js";
import { F as _, f as U } from "./index-kYufzuXV.js";
import { H as q, a as z, b as Q, c as V, d as W, e as K, h as R } from "./index-ZR5YJ3lO.js";
import { L as J, a as X, b as Y, c as Z, d as $, e as aa, f as oa, l as sa } from "./index-LBElx1HN.js";
import { M as ea, a as ra, m as ta } from "./index-DOqPBmZz.js";
import { N as ia, a as na, b as da, c as ma, d as la, e as pa, n as ca } from "./index-Dksp97ji.js";
import { P as fa, a as ua, p as ya } from "./index-C8NlT4sQ.js";
import { P as Ca, a as ba, b as Sa, p as ga } from "./index-dODLOINB.js";
import { Q as va, q as La } from "./index-OptCax2o.js";
import { S as Ea, a as Fa, s as ha } from "./index-BrpAs-2C.js";
import { S as Pa, s as Aa } from "./index-XkoJrs3Q.js";
import { S as wa, s as Ba } from "./index-CSEUxTcU.js";
import { T as Ha, t as Ia } from "./index-CoevyLaX.js";
import { T as Na, t as xa } from "./index-BXNlNViX.js";
const Da = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  AccordionItem: r,
  ActionsDisplay: i,
  AlertPage: d,
  AlertPromo: m,
  AlertSite: l,
  BrandCardStack: c,
  BrandChevronScroll: f,
  CardArticle: y,
  CardEvent: C,
  CardIcon: b,
  CardOverlay: S,
  CardStandard: g,
  CardVideo: v,
  CarouselBase: E,
  CarouselCards: F,
  CarouselImageMultiple: h,
  CarouselImageSingle: P,
  CarouselThumbnail: A,
  CarouselWide: w,
  FeedAlert: H,
  FeedEventsGrid: I,
  FeedEventsGrouped: N,
  FeedEventsList: x,
  FeedExpertsBio: D,
  FeedExpertsGrid: O,
  FeedExpertsList: T,
  FeedNewsFeatured: G,
  FeedNewsGrid: j,
  FeedNewsList: k,
  FooterOptions: _,
  HeroBase: q,
  HeroExpand: z,
  HeroGrid: Q,
  HeroLogo: V,
  HeroMinimal: W,
  HeroVideoArrow: K,
  LayoutBoxLogo: J,
  LayoutImageExpand: X,
  LayoutModal: Y,
  LayoutScrollTop: Z,
  LayoutSectionIntroSmall: $,
  LayoutSectionIntroWide: aa,
  LayoutStickyColumns: oa,
  MediaGif: ea,
  MediaInline: ra,
  NavigationBreadcrumb: ia,
  NavigationDrawer: na,
  NavigationHeader: da,
  NavigationItem: ma,
  NavigationSlider: la,
  NavigationSticky: pa,
  PathwayHighlight: fa,
  PathwayImage: ua,
  PersonBio: Ca,
  PersonDisplay: ba,
  PersonHero: Sa,
  QuoteDisplay: va,
  SliderEventDisplay: Ea,
  SliderEventFeed: Fa,
  SocialSharing: Pa,
  StatDisplay: wa,
  TabDisplay: Ha,
  TextEventLockup: Na,
  accordion: t,
  actions: n,
  alert: p,
  brand: u,
  card: L,
  carousel: B,
  feed: M,
  footer: U,
  hero: R,
  layout: sa,
  media: ta,
  navigation: ca,
  pathway: ya,
  person: ga,
  quote: La,
  slider: ha,
  social: Aa,
  stat: Ba,
  tab: Ia,
  text: xa
}, Symbol.toStringTag, { value: "Module" })), Oa = (a) => {
  for (const s in a)
    if (Object.prototype.hasOwnProperty.call(a, s)) {
      const o = a[s];
      for (const e in o)
        Object.prototype.hasOwnProperty.call(o, e) && o[e]();
    }
}, Ta = Da, so = () => {
  console.error(
    `[DEPRECATED] LoadUmdComponents is deprecated.
Use: import { initializeBundle } from "@universityofmaryland/web-components-library/bundle"`
  ), Oa(Ta);
};
export {
  so as L,
  Oa as l,
  Da as u
};
