// @universityofmaryland/web-styles-library/dist/exports/generate.js
import * as token from "@universityofmaryland/web-token-library";
import * as typography from "../typography.js";
import * as layout from "../layout.js";
import * as element from "../element.js";
import * as animation from "../animation.js";
import * as accessibility from "../accessibility.js";
import { webComponentStyles } from "../web-components.js";
import { reset, root } from "../root.js";
import { removeDuplicates, minifyCSS } from "../utilities/transform/css.js";
import { createBlock, createClassSelector, createRules, formatValue, isPlainObject, toKebabCase } from "../utilities/transform/css.js";
import { processNestedJssObjects, processWebComponentStyles, generateBaseCSS } from "../utilities/transform/jss.js";
import { convertFontToCss, convertToClassSelectorCss, convertToCss, convertToSelectorCSS, formatNestedObjects, formatWithClassSelector } from "../utilities/transform/jss.js";
import { generateTokensCSS } from "../utilities/transform/variables.js";
import { formatTokenKey, fromTokens, toCssObject, toString } from "../utilities/transform/variables.js";
import { base64fonts } from "../typography/font-face/index.js";
async function generateCSSStrings() {
  const tokensCSS = generateTokensCSS(
    token.color,
    token.spacing,
    token.font,
    token.media
  );
  const fontFacesCSS = base64fonts;
  const typographyCSS = processNestedJssObjects(typography);
  const typographyCSSClean = await removeDuplicates(typographyCSS);
  const layoutCSS = processNestedJssObjects(layout);
  const layoutCSSClean = await removeDuplicates(layoutCSS);
  const elementCSS = processNestedJssObjects(element);
  const elementCSSClean = await removeDuplicates(elementCSS);
  const animationCSS = processNestedJssObjects(animation);
  const animationCSSClean = await removeDuplicates(animationCSS);
  const accessibilityCSS = processNestedJssObjects(accessibility);
  const accessibilityCSSClean = await removeDuplicates(accessibilityCSS);
  const webComponentsCSS = processWebComponentStyles(webComponentStyles);
  const webComponentsCSSClean = await removeDuplicates(webComponentsCSS);
  const baseCSS = generateBaseCSS(root, reset);
  const baseCSSClean = await removeDuplicates(baseCSS);
  const fullBundle = [
    tokensCSS,
    fontFacesCSS,
    baseCSS,
    typographyCSS,
    layoutCSS,
    elementCSS,
    animationCSS,
    accessibilityCSS,
    webComponentsCSS
  ].join("\n\n");
  const fullBundleClean = await removeDuplicates(fullBundle);
  return {
    tokens: minifyCSS(tokensCSS),
    fontFaces: minifyCSS(fontFacesCSS),
    typography: minifyCSS(typographyCSSClean),
    layout: minifyCSS(layoutCSSClean),
    element: minifyCSS(elementCSSClean),
    animation: minifyCSS(animationCSSClean),
    accessibility: minifyCSS(accessibilityCSSClean),
    webComponents: minifyCSS(webComponentsCSSClean),
    base: minifyCSS(baseCSSClean),
    styles: minifyCSS(fullBundleClean)
  };
}
export {
  convertFontToCss,
  convertToClassSelectorCss,
  convertToCss,
  convertToSelectorCSS,
  createBlock,
  createClassSelector,
  createRules,
  formatNestedObjects,
  formatTokenKey,
  formatValue,
  formatWithClassSelector,
  fromTokens,
  generateBaseCSS,
  generateCSSStrings,
  generateTokensCSS,
  isPlainObject,
  minifyCSS,
  processNestedJssObjects,
  processWebComponentStyles,
  removeDuplicates,
  toCssObject,
  toKebabCase,
  toString
};
